File: gui/menu/data/edit_tables.cpp
#include "edit_tables.h"


MenuDataEditTables::MenuDataEditTables(QWidget* parent)
    : IodeSettingsDialog(parent)
{
    setupUi(this);

    completer = new IodeCompleter(false, false, TABLES, textEdit_table_names);
    textEdit_table_names->setCompleter(completer);

    wTablesNames = new WrapperQPlainTextEdit(label_table_names->text(), *textEdit_table_names, REQUIRED_FIELD);
    mapFields["TablesNames"] = wTablesNames;

    className = "MENU_DATA_EDIT_TABLES";
    loadSettings();
}

MenuDataEditTables::~MenuDataEditTables()
{
    clear_all_reference_kdbs();
    foreach(ComputedTableNumericalDialog* view, tableViews) view->close();
    tableViews.clear();

    delete wTablesNames;
    delete completer;
}


void MenuDataEditTables::display()
{
    try
    {
        Sample* smpl = Variables.get_sample();
        QString gsample = QString::fromStdString(smpl->start_period().to_string()) + ":" + QString::number(smpl->nb_periods());

        QString tableNames = wTablesNames->extractAndVerify();

        // Parses a string and replaces @filename and $listname by their contents
        char** c_table_names = B_ainit_chk(tableNames.toStdString().data(), NULL, 0);
        int nb_names = SCR_tbl_size((unsigned char**) c_table_names);
        for(int i=0; i < nb_names; i++)
        {
            QString tableName(c_table_names[i]); 
            ComputedTableNumericalDialog* view = new ComputedTableNumericalDialog(tableName, gsample, 4, "", this);
            tableViews.append(view);
            view->open();
        }
    }
    catch(const std::exception& e)
    {
        QMessageBox::critical(this, tr("ERROR"), tr(e.what()));
        return;
    }
}

void MenuDataEditTables::help()
{
	QDesktopServices::openUrl(url_manual);
}

File: gui/menu/data/edit_tables.py
from PySide6.QtCore import Slot
from PySide6.QtWidgets import QWidget, QMessageBox
from PySide6.QtGui import QCloseEvent

from settings import MixinSettingsDialog
from tabs.iode_objs.tab_computed_table import ComputedTableNumericalDialog
from .ui_edit_tables import Ui_MenuDataEditTables

from typing import List
from iode import IodeType, tables, variables, Sample, Table, reset_extra_files
from iode.util import ComputedTable


class MenuDataEditTables(MixinSettingsDialog):
    def __init__(self, parent: QWidget=None):
        super().__init__(parent)
        self.ui = Ui_MenuDataEditTables()
        self.ui.setupUi(self)
        self.prepare_settings(self.ui)

        self.ui.textEdit_table_names.handle_iode_type(IodeType.TABLES)        
        self.ui.textEdit_table_names.include_iode_command(False)        
        self.ui.textEdit_table_names.include_lec_functions(False)  

        self.table_views: List[ComputedTableNumericalDialog] = []
        self.load_settings()

    # override MixinSettingsDialog method
    @Slot()
    def closeEvent(self, event: QCloseEvent):
        self.save_settings()

        for dialog in self.table_views: 
            dialog.close()
        self.table_views.clear()
        reset_extra_files()

        event.accept()

    @Slot()
    def display(self):
        try:
            sample: Sample = variables.sample
            generalized_sample: str = f"{str(sample.start)}:{sample.nb_periods}"
            pattern: str = ";".join(self.ui.textEdit_table_names.toPlainText().splitlines()).strip()

            table_names: List[str] = tables.get_names(pattern)
            for table_name in table_names:
                computed_table_dialog = ComputedTableNumericalDialog(table_name, generalized_sample, 4, parent=self)
                self.table_views.append(computed_table_dialog)
                computed_table_dialog.open()
        except Exception as e:
            QMessageBox.warning(self, "WARNING", f"Failed to edit table(s) .\n" + str(e))
